cdc/puller/multiplexing_puller.go

```diff
@@ -0,0 +1,423 @@
+	"github.com/tikv/client-go/v2/oracle"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	resolveLockFence        time.Duration = 20 * time.Second
+	resolveLockTickInterval time.Duration = 10 * time.Second
+)
+
+type tableProgress struct {
+	changefeed model.ChangeFeedID
+	client     *kv.SharedClient
+	spans      []tablepb.Span
+	startTs    model.Ts
+	tableName  string
+
+	initialized          bool
+	resolvedTsUpdated    time.Time
+	resolvedTs           atomic.Uint64
+	maxIngressResolvedTs atomic.Uint64
+
+	resolvedSpans chan kv.MultiplexingEvent
+	tsTracker     frontier.Frontier
+
+	consume struct {
+		sync.RWMutex
+		removed bool
+		f       func(context.Context, *model.RawKVEntry, []tablepb.Span) error
+	}
+
+	scheduled atomic.Bool
+}
+
+// MultiplexingPuller works with `kv.SharedClient`. All tables share resources.
+type MultiplexingPuller struct {
+	changefeed model.ChangeFeedID
+	client     *kv.SharedClient
+	consume    func(context.Context, *model.RawKVEntry, []tablepb.Span) error
+	workers    int
+	hasher     func(tablepb.Span, int) int
+	frontiers  int
+
+	// inputChs is used to collect events from client.
+	inputChs []chan kv.MultiplexingEvent
+	// advanceCh is used to handle resolved ts in frontier workers.
+	advanceCh chan *tableProgress
+
+	// NOTE: different subscriptions can share one tableProgress.
+	subscriptions struct {
+		sync.RWMutex
+		m map[kv.SubscriptionID]*tableProgress
+		n *spanz.HashMap[*tableProgress]
+	}
+
+	pullerEventCounterKv       prometheus.Counter
+	pullerEventCounterResolved prometheus.Counter
+	queueKvDuration            prometheus.Observer
+	queueResolvedDuration      prometheus.Observer
+}
+
+// NewMultiplexingPuller creates a MultiplexingPuller. Outputs are handled by
+// `consume`, which will be called in several sub-routines concurrently.
+//
+// `workers` specifies how many workers will be spawned to handle events.
+// `frontiers` specifies how many workers will be spawned to handle resolved timestamps.
+func NewMultiplexingPuller(
+	changefeed model.ChangeFeedID,
+	client *kv.SharedClient,
+	consume func(context.Context, *model.RawKVEntry, []tablepb.Span) error,
+	workers int,
+	hasher func(tablepb.Span, int) int,
+	frontiers int,
+) *MultiplexingPuller {
+	x := &MultiplexingPuller{
+		changefeed: changefeed,
+		client:     client,
+		consume:    consume,
+		workers:    workers,
+		hasher:     hasher,
+		frontiers:  frontiers,
+		advanceCh:  make(chan *tableProgress, 128),
+	}
+	x.subscriptions.m = make(map[kv.SubscriptionID]*tableProgress)
+	x.subscriptions.n = spanz.NewHashMap[*tableProgress]()
+
+	x.inputChs = make([]chan kv.MultiplexingEvent, 0, workers)
+	for i := 0; i < workers; i++ {
+		x.inputChs = append(x.inputChs, make(chan kv.MultiplexingEvent, 1024))
+	}
+	return x
+}
+
+// Subscribe some spans. They will share one same resolved timestamp progress.
+func (p *MultiplexingPuller) Subscribe(spans []tablepb.Span, startTs model.Ts, tableName string) []kv.SubscriptionID {
```

Comment:
The return value of `Subscribe` has never been processed by its caller, so would it be more appropriate to change it to a method with no return value?